### ISSUE TITLE
(*COMMIT) is broken by some sort of optimization

### ISSUE THREAD

**Issue by PhilipHazel** [Perl/perl5#18770]
This is a bug report for perl from Philip.Hazel@gmail.com,
generated with the help of perlbug 1.42 running under perl 5.32.1.

-----------------------------------------------------------------
[Please describe your issue here]

Some kind of optimization in Perl is breaking some uses of (*COMMIT). Here are 
two examples. The first matches, when I expect it to fail. The second, where I 
have turned off optimization, does not match, which is what I expected.

$ perl -e 'if (abcdefabcxyz =~ /abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; }'
yes >abcxyz<

$ perl -e 'if (abcdefabcxyz =~ /(??{})abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; }'
no 

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=core
    severity=low
---
Site configuration information for perl 5.32.1:

Configured by builduser at Sat Feb  6 11:58:58 CET 2021.

Summary of my perl5 (revision 5 version 32 subversion 1) configuration:
   
  Platform:
    osname=linux
    osvers=5.10.3-arch1-1
    archname=x86_64-linux-thread-multi
    uname='archlinux'
    config_args='-des -Dusethreads -Duseshrplib -Doptimize=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -Dprefix=/usr -Dvendorprefix=/usr -Dprivlib=/usr/share/perl5/core_perl -Darchlib=/usr/lib/perl5/5.32/core_perl -Dsitelib=/usr/share/perl5/site_perl -Dsitearch=/usr/lib/perl5/5.32/site_perl -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/5.32/vendor_perl -Dscriptdir=/usr/bin/core_perl -Dsitescript=/usr/bin/site_perl -Dvendorscript=/usr/bin/vendor_perl -Dinc_version_list=none -Dman1ext=1perl -Dman3ext=3perl -Dcccdlflags='-fPIC' -Dlddlflags=-shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Dldflags=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Dmyuname=archlinux -Dmyhostname=archlinux -Dcf_time=Sat Feb  6 10:58:54 AM UTC 2021'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef

  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-march=x86-64 -mtune=generic -O2 -pipe -fno-plt'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='10.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/include-fixed /usr/lib /lib/../lib /usr/lib/../lib /lib /lib64 /usr/lib64
    libs=-lpthread -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.33.so
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version='2.3  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.32/core_perl/CORE'
    cccdlflags='-fPIC'
    lddlflags='-shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -fstack-protector-strong'

---
@INC for perl 5.32.1:
    /usr/lib/perl5/5.32/site_perl
    /usr/share/perl5/site_perl
    /usr/lib/perl5/5.32/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/5.32/core_perl
    /usr/share/perl5/core_perl

---
Environment for perl 5.32.1:
    HOME=/home/ph10
    LANG=en_GB.utf8
    LANGUAGE=en_GB.utf8
    LC_ALL=C
    LC_COLLATE=C
    LD_LIBRARY_PATH=/usr/local/lib
    LOGDIR (unset)
    PATH=/home/ph10/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/opt/android-sdk/platform-tools:/opt/android-sdk/tools:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

**Comment by jkeenan** [comment]
On 5/6/21 11:40 AM, PhilipHazel wrote:
> This is a bug report for perl from ***@***.*** 
> ***@***.***>,
> generated with the help of perlbug 1.42 running under perl 5.32.1.
> 
> ------------------------------------------------------------------------
> 
> [Please describe your issue here]
> 
> Some kind of optimization in Perl is breaking some uses of (*COMMIT). 
> Here are
> two examples. The first matches, when I expect it to fail. The second, 
> where I
> have turned off optimization, does not match, which is what I expected.
> 
> $ perl -e 'if (abcdefabcxyz =~ /abc(*COMMIT)xyz/) { print "yes >$&<\n"; 
> } else { print "no \n"; }'
> yes >abcxyz<
> 
> $ perl -e 'if (abcdefabcxyz =~ /(??{})abc(*COMMIT)xyz/) { print "yes 
>  >$&<\n"; } else { print "no \n"; }'
> no
> 

Could you describe what you mean by "turning off [or turning on] 
optimization"?

Does this mean you ran this against differently configured Perls?  If 
so, what differences in config_args did you use?

Thank you very much.
Jim Keenan

**Comment by PhilipHazel** [comment]
Hi,

The two Perl commands shown in my report were run against the same version
of Perl. The patterns are identical, except that I started the second one
with (??{""}) which, I was told some time ago, would cause Perl not to
apply its optimizations. It of course, does nothing as far as the matching
is concerned, as it calls for the execution of an empty bit of Perl code.
So the result of the two matches should be identical.

Regards,
Philip

**Comment by Leont** [comment]
Running both examples with `-Mre=debug` gives some more information:
```
leont@leon-desktop:~/Code/perl$ perl -Mre=debug -e 'if (abcdefabcxyz =~ /abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; }'
Compiling REx "abc(*COMMIT)xyz"
Final program:
   1: EXACT <abc> (3)
   3: COMMIT (5)
   5: EXACT <xyz> (7)
   7: END (0)
anchored "abcxyz" at 0..0 (checking anchored) minlen 6 
Matching REx "abc(*COMMIT)xyz" against "abcdefabcxyz"
Intuit: trying to determine minimum start position...
  doing 'check' fbm scan, [0..12] gave 6
  Found anchored substr "abcxyz" at offset 6 (rx_origin now 6)...
  (multiline anchor test skipped)
  try at offset...
Intuit: Successfully guessed: match at offset 6
   6 <abcdef> <abcxyz>       |   0| 1:EXACT <abc>(3)
   9 <abcdefabc> <xyz>       |   0| 3:COMMIT(5)
   9 <abcdefabc> <xyz>       |   1|  5:EXACT <xyz>(7)
  12 <abcdefabcxyz> <>       |   1|  7:END(0)
Match successful!
yes >abcxyz<
Freeing REx: "abc(*COMMIT)xyz"
leont@leon-desktop:~/Code/perl$ perl -Mre=debug -e 'if (abcdefabcxyz =~ /(??{})abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; }'
Compiling REx "(??{})abc(*COMMIT)xyz"
Final program:
   1: LOGICAL[2] (2)
   2: EVAL (5)
   5: EXACT <abc> (7)
   7: COMMIT (9)
   9: EXACT <xyz> (11)
  11: END (0)
floating "abcxyz" at 0..9223372036854775807 (checking floating) minlen 6 with eval 
Matching REx "(??{})abc(*COMMIT)xyz" against "abcdefabcxyz"
Intuit: trying to determine minimum start position...
  doing 'check' fbm scan, [0..12] gave 6
  Found floating substr "abcxyz" at offset 6 (rx_origin now 0)...
  (multiline anchor test skipped)
Intuit: Successfully guessed: match at offset 0
   0 <> <abcdefabcx>         |   0| 1:LOGICAL[2](2)
   0 <> <abcdefabcx>         |   0| 2:EVAL(5)
Compiling REx ""
Final program:
   1: NOTHING (2)
   2: END (0)
minlen 0 
EVAL/GOSUB: Matching embedded REx "" against "abcdefabcxyz"
   0 <> <abcdefabcx>         |   1|  1:NOTHING(2)
   0 <> <abcdefabcx>         |   1|  2:END(0)
                             |   1|  END: EVAL trying tail ... (cur_eval=0)
   0 <> <abcdefabcx>         |   2|   5:EXACT <abc>(7)
   3 <abc> <defabcxyz>       |   2|   7:COMMIT(9)
   3 <abc> <defabcxyz>       |   3|    9:EXACT <xyz>(11)
                             |   3|    failed...
                             |   2|   failed...
Freeing REx: ""
Match failed
no 
```

It seems like the optimized code is looking for `abcxyz`, when it should be lookign for `abc`

**Comment by khwilliamson** [comment]
On 5/6/21 10:06 AM, Leon Timmermans wrote:
> Running both examples with |-Mre=debug| gives some more information:
> 
> ***@***.***:~/Code/perl$ perl -Mre=debug -e 'if (abcdefabcxyz =~ 
> /abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; }' 
> Compiling REx "abc(*COMMIT)xyz" Final program: 1: EXACT <abc> (3) 3: 
> COMMIT (5) 5: EXACT <xyz> (7) 7: END (0) anchored "abcxyz" at 0..0 
> (checking anchored) minlen 6 Matching REx "abc(*COMMIT)xyz" against 
> "abcdefabcxyz" Intuit: trying to determine minimum start position... 
> doing 'check' fbm scan, [0..12] gave 6 Found anchored substr "abcxyz" at 
> offset 6 (rx_origin now 6)... (multiline anchor test skipped) try at 
> offset... Intuit: Successfully guessed: match at offset 6 6 <abcdef> 
> <abcxyz> | 0| 1:EXACT <abc>(3) 9 <abcdefabc> <xyz> | 0| 3:COMMIT(5) 9 
> <abcdefabc> <xyz> | 1| 5:EXACT <xyz>(7) 12 <abcdefabcxyz> <> | 1| 
> 7:END(0) Match successful! yes >abcxyz< Freeing REx: "abc(*COMMIT)xyz" 
> ***@***.***:~/Code/perl$ perl -Mre=debug -e 'if (abcdefabcxyz =~ 
> /(??{})abc(*COMMIT)xyz/) { print "yes >$&<\n"; } else { print "no \n"; 
> }' Compiling REx "(??{})abc(*COMMIT)xyz" Final program: 1: LOGICAL[2] 
> (2) 2: EVAL (5) 5: EXACT <abc> (7) 7: COMMIT (9) 9: EXACT <xyz> (11) 11: 
> END (0) floating "abcxyz" at 0..9223372036854775807 (checking floating) 
> minlen 6 with eval Matching REx "(??{})abc(*COMMIT)xyz" against 
> "abcdefabcxyz" Intuit: trying to determine minimum start position... 
> doing 'check' fbm scan, [0..12] gave 6 Found floating substr "abcxyz" at 
> offset 6 (rx_origin now 0)... (multiline anchor test skipped) Intuit: 
> Successfully guessed: match at offset 0 0 <> <abcdefabcx> | 0| 
> 1:LOGICAL[2](2) 0 <> <abcdefabcx> | 0| 2:EVAL(5) Compiling REx "" Final 
> program: 1: NOTHING (2) 2: END (0) minlen 0 EVAL/GOSUB: Matching 
> embedded REx "" against "abcdefabcxyz" 0 <> <abcdefabcx> | 1| 
> 1:NOTHING(2) 0 <> <abcdefabcx> | 1| 2:END(0) | 1| END: EVAL trying tail 
> ... (cur_eval=0) 0 <> <abcdefabcx> | 2| 5:EXACT <abc>(7) 3 <abc> 
> <defabcxyz> | 2| 7:COMMIT(9) 3 <abc> <defabcxyz> | 3| 9:EXACT <xyz>(11) 
> | 3| failed... | 2| failed... Freeing REx: "" Match failed no |
> 
> It seems the optimizer is looking for |abcxyz|, when it should be 
> lookign for |abc|
> 
> —

FWIW, I tried bisecting this, and this is broken for all Perl official 
releases

**Comment by khwilliamson** [comment]
The syntax was added by
commit e2e6a0f1870d05ddb1ce18fd8556b71330dc694c
 Author: Yves Orton <demerphq@gmail.com>
 Date:   Mon Nov 6 14:06:28 2006 +0100
 
     New regex syntax omnibus
     Message-ID: <9b18b3110611060406u2fa1572as57073949a5df9e62@mail.gmail.com>
     
     Plus a portability fix (in string comparison for regex verbs)
     and doc tweaks / podchecker fixes

@demerphq any comments?

**Comment by hvds** [comment]
Patch below should fix it; I'll clean it up and add tests after blead is unfrozen. There are probably more of those verbs that should do this.

```
--- a/regcomp.c
+++ b/regcomp.c
@@ -6527,6 +6527,9 @@ Perl_re_printf( aTHX_  "LHS=%" UVuf " RHS=%" UVuf "\n",
 
 #endif /* old or new */
 #endif /* TRIE_STUDY_OPT */
+        else if (OP(scan) == COMMIT) {
+            flags &= ~SCF_DO_SUBSTR;
+        }
 
        /* Else: zero-length, ignore. */
        scan = regnext(scan);
```

**Comment by hvds** [comment]
Looks like the change for *COMMIT also needs a scan_commit() call; *ACCEPT already turns off SCF_DO_SUBSTR, but also needs to turn off SCF_DO_STCLASS. Since they are at least a little related, I've moved the new *COMMIT stanza next to where *ACCEPT is handled in [branch gh18770](https://github.com/Perl/perl5/tree/gh18770).

**Comment by hvds** [comment]
I've rebased the branch and opened #18809 for review.

**Comment by hvds** [comment]
PR #18809 is now merged as 499aa13271; I'll close this after I see some smoke reports.
